wrapper/src/main/java/software/amazon/jdbc/plugin/bluegreen/BlueGreenConnectionPlugin.java

```diff
@@ -0,0 +1,321 @@
+    methods.add("connect");
+    methods.addAll(SubscribedMethodHelper.NETWORK_BOUND_METHODS);
+    this.subscribedMethods = Collections.unmodifiableSet(methods);
+  }
+
+  @Override
+  public Set<String> getSubscribedMethods() {
+    return this.subscribedMethods;
+  }
+
+  @Override
+  public Connection connect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> connectFunc)
+      throws SQLException {
+
+    this.resetHoldTimeNano();
+
+    try {
+
+      this.bgStatus = this.pluginService.getStatus(BlueGreenStatus.class, this.bgdId);
+
+      if (this.bgStatus == null) {
+        return regularOpenConnection(connectFunc, isInitialConnection);
+      }
+
+      if (isInitialConnection) {
+        this.isIamInUse = this.pluginService.isPluginInUse(IamAuthConnectionPlugin.class);
+      }
+
+      BlueGreenRole hostRole = this.bgStatus.getRole(hostSpec);
+
+      if (hostRole == null) {
+        // Connection to a host that isn't participating in BG switchover.
+        return regularOpenConnection(connectFunc, isInitialConnection);
+      }
+
+      Connection conn = null;
+      ConnectRouting routing = this.bgStatus.getConnectRouting().stream()
+          .filter(r -> r.isMatch(hostSpec, hostRole))
+          .findFirst()
+          .orElse(null);
+
+      if (routing != null) {
+        this.startTimeNano.set(this.getNanoTime());
+      }
+
+      while (routing != null && conn == null) {
+        conn = routing.apply(
+            this,
+            hostSpec,
+            props,
+            isInitialConnection,
+            connectFunc,
+            this.pluginService);
+
+        this.endTimeNano.set(this.getNanoTime());
+
+        if (conn == null) {
+
+          this.bgStatus = this.pluginService.getStatus(BlueGreenStatus.class, this.bgdId);
+
+          routing = this.bgStatus.getConnectRouting().stream()
+              .filter(r -> r.isMatch(hostSpec, hostRole))
+              .findFirst()
+              .orElse(null);
+        }
+      }
+
+      if (conn == null) {
+        this.endTimeNano.set(this.getNanoTime());
+        conn = connectFunc.call();
+      }
+
+      if (isInitialConnection) {
+        // Provider should be initialized after connection is open and a dialect is properly identified.
+        this.initProvider();
+      }
+
+      return conn;
+
+    } finally {
+      if (this.startTimeNano.get() > 0) {
+        this.endTimeNano.compareAndSet(0, this.getNanoTime());
+      }
+    }
+  }
+
+  protected Connection regularOpenConnection(
+      final JdbcCallable<Connection, SQLException> connectFunc,
+      final boolean isInitialConnection) throws SQLException {
+
+    Connection conn = connectFunc.call();
+
+    if (isInitialConnection) {
+      // Provider should be initialized after connection is open and a dialect is properly identified.
+      this.initProvider();
+    }
```

Comment:
I think this `isInitialConnection` check is redundant.
In `initProvider` we call `provider.computeIfAbsent`, so this call would be no-op after the initial connection anyway.

Reply:
Checking `isInitialConnection` is less expensive than checking provider map.